[CBR-464] Raise Correct `CoinSelHardErr` during coin selection

KtorZ opened the pull request:
## Description

In a previous PR (#3704 - CBR-462), we patched the coin selection and remap the `UTxODepleted` error to a `CannotCoverFee` error. 
From this, the frontend team discovered a new bug caused because again, `UTxODepleted` was thrown instead of a correct `UTxOExhausted`. 

Instead of patching again and re-mapping the error to the correct one, I took a step back and slightly reviewed the original abstraction such that we would now have:

```hs
type PickUtxo m = Core.Coin  -- ^ Fee to still cover
               -> CoinSelT Core.Utxo CoinSelHardErr m (Maybe (Core.TxIn, Core.TxOutAux))
```

instead of

```hs
type PickUtxo m = Core.Coin  -- ^ Fee to still cover
               -> CoinSelT Core.Utxo CoinSelHardErr m (Core.TxIn, Core.TxOutAux)
```

The rational here is that, the outcome may depend of the context. We use this function in
multiple places where running out of UTxO may have a different meaning. Therefore, instead
of introducing the 'UtxoDepleted' error, we return a raw Maybe and let the caller decides
what error should be thrown (here, most likely: CannotCoverFee or UtxoExhausted)


---

Note that in theory, we could simply have the following to make it obvious that `PickUtxo` _can't fail_:

```hs
type PickUtxo m = Core.Coin  -- ^ Fee to still cover
               -> StrictStateT Core.Utxo m (Maybe (Core.TxIn, Core.TxOutAux))

-- | Hoist (m ~> ExceptT e m) to handle non failing implementation
hoistExceptT :: Functor m => StrictStateT utxo m a -> CoinSelT utxo e m a
hoistExceptT = CoinSelT . mapStrictStateT (ExceptT . fmap Right)
```

However, this exposes the `StrictStateT` from `CoinSelT` which is opaque.. So, we could go even further and have exposed instead. 

```hs
newtype NonFailingCoinSelT utxo m a :: NonFailingCoinSelT (StrictStateT utxo m a)
  deriving ( Functor
           , Applicative
           , Monad
           , MonadState utxo
           )
```

## Linked issue

[[CBR-464]](https://iohk.myjetbrains.com/youtrack/issue/CBR-464)

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] ~~CHANGELOG entry has been added and is linked to the correct PR on GitHub.~~

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

KtorZ commented:
```
fails if you don't have any money 
    ClientWalletError (NotEnoughMoney (ErrAvailableBalanceIsInsufficient 0))
    +++ OK, passed 1 tests.

fails if you spend more money than your available balance
    ClientWalletError (NotEnoughMoney (ErrAvailableBalanceIsInsufficient 42))
    +++ OK, passed 1 tests.

fails if you can't cover fee with a transaction
    ClientWalletError (NotEnoughMoney ErrCannotCoverFee)
    +++ OK, passed 1 tests.
```

KtorZ commented on wallet-new/src/Cardano/Wallet/API/V1/ReifyWalletError.hs:
```diff
@@ -267,15 +268,13 @@ newTransactionError e = case e of
           ex@(CoinSelHardErrUtxoExhausted balance _payment) ->
-              case (readMaybe $ T.unpack balance) of
+              -- NOTE balance & payment are "prettified" coins representation (e.g. "42 coin(s)")
+              case (readMaybe $ T.unpack $ T.dropWhileEnd (not . C.isDigit) balance) of
```
That's what I understood from this indeed. Any arguments about having `Coin` here instead of `Int` since it conveys more meaning. I don't see any reasons why we would do an early conversion when throwing the error.

edsko replied:
`Coin` is not applicable, because that is specific to the Cardano domain. Other domains will not use `Coin` for their representation of values.

KtorZ replied:
Ho true, forgot that the errors were generics here :+1:

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:
```diff
@@ -288,7 +288,6 @@ runCoinSelT opts pickUtxo policy request utxo = do
     policy' :: CoinSelT Core.Utxo CoinSelHardErr m
                  ([CoinSelResult Cardano], SelectedUtxo Cardano)
     policy' = do
-        when (Map.null utxo) $ throwError CoinSelHardErrUtxoDepleted
```
Me neither :upside_down_face:  ... Hence the need for this PR. 
The previous patches felt a bit like "monkey-patching" without really tackling the actual issue.

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:
```diff
@@ -387,10 +386,10 @@ largestFirst opts maxInps =
         search :: [(Core.TxIn, Core.TxOutAux)]
-               -> CoinSelT Core.Utxo CoinSelHardErr m (Core.TxIn, Core.TxOutAux)
+               -> CoinSelT Core.Utxo CoinSelHardErr m (Maybe (Core.TxIn, Core.TxOutAux))
         search [] = throwError CoinSelHardErrCannotCoverFee
```
Indeed

KtorZ commented:
@edsko @adinapoli-iohk 
- [x] Addressed most comments you had on this PR and the two related others. Divvy fees now just throws if the second invariant is broken (a zero output is given) and enforced by the callers.

- [x] I now pass around the utxo's balance instead of the complete utxo to construct a more meaningful error. 

- [x] Tweaked a bit the type for `PickUtxo` to make it clear that it cannot throw (with an appropriate note).

- [x] Opened a ticket to fix the `updateForeign` / `updatePending` calls and have the foreign transaction appears in the pre-filtered input list (in which case we can simply ditch `updateForeign` entirely).

---

I didn't fix the parsing of the balance from the `CoinSelHarderrUtxoExhausted`

```hs
          ex@(CoinSelHardErrUtxoExhausted balance _payment) ->
              case (readMaybe $ T.unpack balance) of
              -- NOTE balance & payment are "prettified" coins representation (e.g. "42 coin(s)")
              case (readMaybe $ T.unpack $ T.dropWhileEnd (not . C.isDigit) balance) of
```

I'll open a ticket for this as it requires a bit more change and is quite isolated from this particular PR.

adinapoli-iohk commented on wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs:
```diff
@@ -150,14 +144,22 @@ feeFromChange totalFee =
   where
+    -- divvyFee only accepts non-zero output values as it doesn't make sense
+    -- to apply fees to those.
+    (cs0, cs') = partition (== valueZero) cs
```
Do we need this `partition`, at all? With @edsko we initially thought a simple `filter` for `(> valueZero)` might have been enough. What are we missing?

KtorZ replied:
As far as I understand, we still want to return those zero value as part of the change, hence the change from 

```hs
 bimap identity unsafeFeeSum
```

to 

```hs
 bimap (++ cs0) unsafeFeeSum
```

We just don't run the `divvyFee` repartition function on them... I mean, we would filter out the zero value at this stage, but I'd expect this function to return the same number of inputs I gave it, no ?

adinapoli-iohk commented on wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic.hs:
```diff
@@ -592,28 +582,18 @@ nLargestFromListBy f n = \xs ->
-divvyFee f fee [a] | f a == valueZero = [(fee, a)]
-divvyFee f fee as                     = map (\a -> (feeForOut a, a)) as
+divvyFee _ _   []                             = error "divvyFee: empty list"
+divvyFee f _ as | any ((== valueZero) . f) as = error "divvyFee: some outputs are null"
```
Do we still need this special guard? I suspect that now that we pushed the main logic fix inside `reduceChangeOutputs` , we can try to be more liberal here?

KtorZ replied:
Well, this is an invariant we want to enforce. If we don't have that, then we might get a division by zero silently coerced to `Word64` and turned to `0`. 

```hs
λ> ceiling (0 * (14 / 0 :: Double)) :: Word64
```

So, I'll better leave a guard that explicitly fails and tell us that we did something wrong :+1: (like it seems to happen in the unit test :sweat_smile: )

KtorZ commented:
Okay, fixed the `genPayees` generator which was generating null payments :+1:

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/DB/Spec/Read.hs:
```diff
@@ -56,9 +56,7 @@ cpAvailableBalance :: IsCheckpoint c => c -> Core.Coin
-    pendingIns   = Set.union
-        (Pending.txIns $ c ^. cpPending)
-        (Pending.txIns $ c ^. cpForeign)
+    pendingIns   = Pending.txIns $ c ^. cpPending
```
Reverting from #3672

KtorZ commented:
Alles Goed! Not waiting for Darwin here :+1:
